test/Fencer/Rules/Test.hs

```diff
@@ -71,6 +81,78 @@ test_loadRulesRecursively =
         (sortOn domainDefinitionId [domain1, domain2])
         (sortOn domainDefinitionId definitions)
 
+-- | Test that a rule limit unit change adds a new counter and leaves
+-- the old one intact.
```

Comment:
I have several suggestions regarding this test.

First of all, it does not test the actual logic used by the server. It does the lookup on its own instead of using `getLimit`. To make it more faithful, I would propose:

* moving `shouldRateLimitDescriptor` to `Fencer.AppState` and renaming it
* perhaps also renaming `Fencer.AppState` to `Fencer.Logic` or something
* removing the export of `recordHits` and not using `recordHits` in the test

Then the test would also get moved to `Fencer.Logic.Test` since it's not related to the `Fencer.Rules` module.

Secondly, here's what this test does:

```
  load definitions1
  hit counterKey1

  load definitions2
  assert (counterKey1 hasn't changed after 'load definitions2')
  assert (counterKey2 doesn't exist after 'load definitions2')
```

This does not match the description of the test ("Test that a rule limit unit change adds a new counter [...]"). Also, there are bad implementations it does not catch – for instance, what if counters are removed during `recordHits` and not during `loadRules`?

Here's a scenario I would propose instead:

```
  // change the limit to e.g. 4 instead of 400000

  load definitions1
  status <- hit "generic_key/dream11_order_create"
  assert (counterRemainingLimit status == 3)

  load definitions2
  hit "generic_key/dream11_order_create"
  assert (counterRemainingLimit status == 3)  // unaffected by the previous 'hit'

  load definitions1
  hit "generic_key/dream11_order_create"
  assert (counterRemainingLimit status == 2)  // the old counter has persisted!
```